File: e2e/candidate-result.mjs
import { constants } from 'node:fs';
import { open } from 'node:fs/promises';
import contract from './candidate-result-contract.json' with { type: 'json' };
import { check } from './errors.cjs';
export { contract as RESULT_CONTRACT };
export const TABLES = ['auth', 'users', 'notes', 'user_tags'];
export const unknownCounts = () => Object.fromEntries(TABLES.map(t => [t, null]));
export const cleanupState = counts => TABLES.some(t => Number.isSafeInteger(counts[t]) && counts[t] > 0)
  ? 'RESIDUAL' : TABLES.every(t => counts[t] === 0) ? 'PROVEN_ZERO' : 'UNPROVEN';
const keys = (value, expected) => value && typeof value === 'object' && !Array.isArray(value) &&
  JSON.stringify(Object.keys(value).sort()) === JSON.stringify([...expected].sort());

export function resultBinding(m, manifestHash) {
  check(m.e2eIdentityVersion === 2 && /^cd-[1-9][0-9]*-[1-9][0-9]*$/.test(m.candidateId) &&
    m.candidateId === `cd-${m.run.id}-${m.run.attempt}` && /^[a-f0-9]{40}$/.test(m.sourceSha) &&
    /^[a-f0-9]{64}$/.test(manifestHash), 'RESULT_ENVELOPE_INVALID');
  return { version: contract.version, candidateId: m.candidateId, sourceSha: m.sourceSha,
    githubRunId: m.run.id, githubRunAttempt: m.run.attempt, manifestHash };
}

export function initialResult(m, manifestHash) {
  return { ...resultBinding(m, manifestHash), scenario: 'NOT_RUN',
    steps: contract.steps.map(name => ({ name, result: 'NOT_RUN' })), cleanup: unknownCounts(),
    cleanupState: 'UNPROVEN', localReceiptState: 'UNKNOWN', evidenceState: 'UNKNOWN',
    httpsCookieVerified: false, failureCode: null, failureOperation: null,
    recoveryHandle: { type: 'DETERMINISTIC_CANDIDATE_ID', candidateId: m.candidateId } };
}

export function setFailure(result, code) {
  check(Object.hasOwn(contract.failures, code), 'RESULT_ENVELOPE_INVALID');
  result.failureCode = code; result.failureOperation = contract.failures[code];
}

export function validateResult(r, m, manifestHash) {
  const expected = initialResult(m, manifestHash);
  check(keys(r, Object.keys(expected)), 'RESULT_ENVELOPE_INVALID');
  for (const [key, value] of Object.entries(resultBinding(m, manifestHash))) {
    check(r[key] === value, 'RESULT_ENVELOPE_INVALID');
  }
  check(['PASS', 'FAIL', 'NOT_RUN', 'UNKNOWN'].includes(r.scenario) &&
    Array.isArray(r.steps) && r.steps.length === contract.steps.length && r.steps.every((s, i) =>
      keys(s, ['name', 'result']) && s.name === contract.steps[i] && ['PASS', 'FAIL', 'NOT_RUN'].includes(s.result)),
  'RESULT_ENVELOPE_INVALID');
  check(r.scenario !== 'PASS' || r.steps.every(s => s.result === 'PASS'), 'RESULT_ENVELOPE_INVALID');
  check(r.scenario !== 'NOT_RUN' || r.steps.every(s => s.result === 'NOT_RUN'), 'RESULT_ENVELOPE_INVALID');
  check(keys(r.cleanup, TABLES) && TABLES.every(t => r.cleanup[t] === null ||
    Number.isSafeInteger(r.cleanup[t]) && r.cleanup[t] >= 0) && r.cleanupState === cleanupState(r.cleanup),
  'RESULT_ENVELOPE_INVALID');
  check(['UNKNOWN', 'PERSISTED', 'FAILED', 'NOT_REQUIRED'].includes(r.localReceiptState) &&
    ['UNKNOWN', 'PASS', 'FAIL'].includes(r.evidenceState) && typeof r.httpsCookieVerified === 'boolean' &&
    (!r.httpsCookieVerified || r.steps[0].result === 'PASS'), 'RESULT_ENVELOPE_INVALID');
  check(r.failureCode === null ? r.failureOperation === null :
    typeof r.failureCode === 'string' && Object.hasOwn(contract.failures, r.failureCode) && r.failureOperation === contract.failures[r.failureCode],
  'RESULT_ENVELOPE_INVALID');
  check(keys(r.recoveryHandle, ['type', 'candidateId']) &&
    r.recoveryHandle.type === 'DETERMINISTIC_CANDIDATE_ID' && r.recoveryHandle.candidateId === m.candidateId,
  'RESULT_ENVELOPE_INVALID');
  check(Buffer.byteLength(JSON.stringify(r)) <= contract.maxBytes, 'RESULT_ENVELOPE_INVALID');
  return r;
}

export function resultExitCode(r) {
  if (r.cleanupState !== 'PROVEN_ZERO') return 21;
  if (r.localReceiptState === 'FAILED' || r.evidenceState !== 'PASS') return 22;
  if (r.scenario !== 'PASS') return 20;
  return r.failureCode === null && r.httpsCookieVerified ? 0 : 22;
}

export async function writeResult(filename, result, m, manifestHash) {
  validateResult(result, m, manifestHash);
  const file = await open(filename, constants.O_WRONLY | constants.O_NOFOLLOW);
  try {
    const s = await file.stat();
    check(s.isFile() && s.nlink === 1 && s.uid === process.getuid() && (s.mode & 0o777) === 0o600,
      'RESULT_ENVELOPE_INVALID');
    await file.truncate(0);
    await file.writeFile(JSON.stringify(result));
    await file.sync();
  } finally { await file.close(); }
}

File: .github/scripts/candidate_result.py
"""Strict public result boundary. No free-form child text is publishable."""
import json
import os
from pathlib import Path
import stat

import wif_submission as proof

CONTRACT = json.loads((Path(__file__).resolve().parents[2] / 'e2e/candidate-result-contract.json').read_text())
TABLES = ('auth', 'users', 'notes', 'user_tags')


def initial(m, manifest_hash):
    proof.require(m.get('e2eIdentityVersion') == 2 and proof.matches(r'[a-f0-9]{64}', manifest_hash),
                  'RESULT_ENVELOPE_INVALID')
    return {'version': 1, 'candidateId': m['candidateId'], 'sourceSha': m['sourceSha'],
            'githubRunId': m['run']['id'], 'githubRunAttempt': m['run']['attempt'],
            'manifestHash': manifest_hash, 'scenario': 'UNKNOWN',
            'steps': [{'name': n, 'result': 'NOT_RUN'} for n in CONTRACT['steps']],
            'cleanup': dict.fromkeys(TABLES), 'cleanupState': 'UNPROVEN',
            'localReceiptState': 'UNKNOWN', 'evidenceState': 'UNKNOWN', 'httpsCookieVerified': False,
            'failureCode': None, 'failureOperation': None,
            'recoveryHandle': {'type': 'DETERMINISTIC_CANDIDATE_ID', 'candidateId': m['candidateId']}}


def state(counts):
    if any(type(counts[t]) is int and counts[t] > 0 for t in TABLES):
        return 'RESIDUAL'
    return 'PROVEN_ZERO' if all(type(counts[t]) is int and counts[t] == 0 for t in TABLES) else 'UNPROVEN'


def keys(value, expected):
    return type(value) is dict and set(value) == set(expected)


def validate(r, m, manifest_hash):
    def need(ok):
        proof.require(ok, 'RESULT_ENVELOPE_INVALID')
    base = initial(m, manifest_hash)
    need(keys(r, base))
    for key in ('version', 'candidateId', 'sourceSha', 'githubRunId', 'githubRunAttempt', 'manifestHash'):
        need(type(r[key]) is type(base[key]) and r[key] == base[key])
    need(r['scenario'] in ('PASS', 'FAIL', 'NOT_RUN', 'UNKNOWN'))
    need(type(r['steps']) is list and len(r['steps']) == len(CONTRACT['steps']))
    for row, name in zip(r['steps'], CONTRACT['steps']):
        need(keys(row, ('name', 'result')) and row['name'] == name and row['result'] in ('PASS', 'FAIL', 'NOT_RUN'))
    need(r['scenario'] != 'PASS' or all(s['result'] == 'PASS' for s in r['steps']))
    need(r['scenario'] != 'NOT_RUN' or all(s['result'] == 'NOT_RUN' for s in r['steps']))
    need(keys(r['cleanup'], TABLES))
    need(all(v is None or type(v) is int and 0 <= v <= 9007199254740991 for v in r['cleanup'].values()))
    need(r['cleanupState'] == state(r['cleanup']))
    need(r['localReceiptState'] in ('UNKNOWN', 'PERSISTED', 'FAILED', 'NOT_REQUIRED'))
    need(r['evidenceState'] in ('UNKNOWN', 'PASS', 'FAIL'))
    need(type(r['httpsCookieVerified']) is bool and
         (not r['httpsCookieVerified'] or r['steps'][0]['result'] == 'PASS'))
    code = r['failureCode']
    need((code is None and r['failureOperation'] is None) or
         type(code) is str and code in CONTRACT['failures'] and r['failureOperation'] == CONTRACT['failures'][code])
    need(keys(r['recoveryHandle'], ('type', 'candidateId')) and r['recoveryHandle'] == base['recoveryHandle'])
    need(len(json.dumps(r).encode()) <= CONTRACT['maxBytes'])
    return r


def set_failure(r, code):
    proof.require(code in CONTRACT['failures'], 'RESULT_ENVELOPE_INVALID')
    r['failureCode'], r['failureOperation'] = code, CONTRACT['failures'][code]


def exit_code(r):
    if r['cleanupState'] != 'PROVEN_ZERO':
        return 21
    if r['localReceiptState'] == 'FAILED' or r['evidenceState'] != 'PASS':
        return 22
    if r['scenario'] != 'PASS':
        return 20
    return 0 if r['failureCode'] is None and r['httpsCookieVerified'] else 22


def read(filename, expected_stat, m, manifest_hash):
    def pairs(rows):
        out = {}
        for k, v in rows:
            proof.require(k not in out, 'RESULT_ENVELOPE_INVALID')
            out[k] = v
        return out
    try:
        with os.fdopen(os.open(filename, os.O_RDONLY | os.O_NOFOLLOW | os.O_NONBLOCK), 'rb') as file:
            s = os.fstat(file.fileno())
            proof.require(stat.S_ISREG(s.st_mode) and s.st_nlink == 1 and s.st_uid == os.getuid() and
                          stat.S_IMODE(s.st_mode) == 0o600 and
                          (s.st_dev, s.st_ino) == (expected_stat.st_dev, expected_stat.st_ino) and
                          s.st_size <= CONTRACT['maxBytes'], 'RESULT_ENVELOPE_INVALID')
            raw = file.read(CONTRACT['maxBytes'] + 1)
        proof.require(bool(raw), 'RESULT_ENVELOPE_MISSING')
        proof.require(len(raw) <= CONTRACT['maxBytes'], 'RESULT_ENVELOPE_INVALID')
        return validate(json.loads(raw, object_pairs_hook=pairs), m, manifest_hash)
    except FileNotFoundError:
        raise proof.GateError('RESULT_ENVELOPE_MISSING') from None
    except Exception as error:
        code = 'RESULT_ENVELOPE_MISSING' if isinstance(error, proof.GateError) and str(error) == 'RESULT_ENVELOPE_MISSING' else 'RESULT_ENVELOPE_INVALID'
        raise proof.GateError(code) from None


def publish(r, m, manifest_hash, env):
    validate(r, m, manifest_hash)
    safe = json.dumps(r, sort_keys=True, separators=(',', ':'))
    print('CD-C3 E2E result: ' + safe, flush=True)
    if env.get('GITHUB_STEP_SUMMARY'):
        with Path(env['GITHUB_STEP_SUMMARY']).open('a') as file:
            file.write('## Candidate E2E result\n\n```json\n' + safe + '\n```\n')
